### PR TITLE
support configuring workspace deploymentStrategy in DWOC

### PR DESCRIPTION
**Pull request by AObuchow**
### What does this PR do?

This PR adds a new DWOC field `config.workspace.deploymentStrategy` which allows for specifying the workspace deployment strategy in the DWOC. The possible choices are `Recreate` (Default) and `RollingUpdate`.

### What issues does this PR fix or reference?
https://github.com/devfile/devworkspace-operator/issues/1045 & https://github.com/eclipse/che/issues/21974 

### Is it tested? How?

Edit the DWOC and set the `deploymentStrategy` field to `RollingUpdate`. Also, ensure that the only accepted options for the `deploymentStrategy` field are `Recreate` and `RollingUpdate`.

```diff
kind: DevWorkspaceOperatorConfig
apiVersion: controller.devfile.io/v1alpha1
config:
  routing:
    clusterHostSuffix: 192.168.39.117.nip.io
    defaultRoutingClass: basic
  workspace:
+    deploymentStrategy: RollingUpdate
    imagePullPolicy: Always
```

Deploy a workspace and modify a field that would change the deployment spec, for example, modify one of the container's CPU or Memory request. 

To try this out, you could apply the following devworkspace:

```YAML
cat <<EOF | kubectl apply -f - 
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: plain-devworkspace
  namespace: devworkspace-controller
spec:
  started: true
  routingClass: 'basic'
  template:
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryLimit: 5Gi
          memoryRequest: 256Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
EOF
```

Next, watch the workspace status (as it will change when the rolling update occurs): `kubectl get dw -n devworkspace-controller`

Once the workspace is running, you could then increase the `memoryRequest`:  `kubectl edit dwoc -n plain-devworkspace -n devworkspace-controller`

```diff
apiVersion: workspace.devfile.io/v1alpha2
kind: DevWorkspace
(...)
spec:
  routingClass: basic
  started: true
  template:
    components:
    - container:
        command:
        - tail
        - -f
        - /dev/null
        image: quay.io/wto/web-terminal-tooling:next
        memoryLimit: 5Gi
+        memoryRequest: 1Gi
        mountSources: true
        sourceMapping: /projects
      name: web-terminal
```

Under the workspace's deployment (which could be checked on the OpenShift or Kubernetes dashboard) you should see a second pod get created and started before the old one is terminated. 

You should also see the devworkspace status phase change from Running -> Starting -> Running:

```
$ kubectl get dw -n $NAMESPACE -w
NAME                 DEVWORKSPACE ID             PHASE      INFO
plain-devworkspace   workspacec3a273cb552340a5   Starting   Waiting for workspace deployment
plain-devworkspace   workspacec3a273cb552340a5   Running    Workspace is running
plain-devworkspace   workspacec3a273cb552340a5   Running    Workspace is running
plain-devworkspace   workspacec3a273cb552340a5   Running    Workspace is running
plain-devworkspace   workspacec3a273cb552340a5   Starting   Networking ready
plain-devworkspace   workspacec3a273cb552340a5   Starting   Waiting for workspace deployment
plain-devworkspace   workspacec3a273cb552340a5   Running    Workspace is running
```

#### Extra test case: removing the deploymentStrategy from the DWOC

1. Set the `deploymentStrategy` to `RollingUpdate` in the DWOC
2. Apply a workspace (`kubectl apply -f ./samples/plain-devworkspace -n $NAMESPACE`), ensure the deployments `spec.strategy.type` is `RollingUpdate` 
3. Remove the `deploymentStrategy` field from the DWOC
4. Trigger a reconcile on the workspace (e.g. add an annotation)
5. Check the workspace deployment, the `spec.strategy.type` field should be `Recreate`

In other words, when the `deploymentStrategy` field is absent from the DWOC, the deployment strategy will be `Recreate`

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

File: pkg/provision/workspace/deployment.go [hunk 1]
```diff
@@ -214,9 +229,10 @@ func getSpecDeployment(
 					constants.DevWorkspaceIDLabel: workspace.Status.DevWorkspaceId,
 				},
 			},
-			Strategy: appsv1.DeploymentStrategy{
-				Type: appsv1.RecreateDeploymentStrategyType,
-			},
+			Strategy:                deploymentStrategy,
+			MinReadySeconds:         0,
+			RevisionHistoryLimit:    pointer.Int32(10),
+			ProgressDeadlineSeconds: pointer.Int32(600),
```
Comment (amisevsk) [inline pkg/provision/workspace/deployment.go:235]: I get that these are defaults but if we're setting them directly it may be worth considering whether these values are appropriate for our use case. 10 minutes for progress is a very long time, and we don't necessarily need 10 revisions in the history (and I believe this setting would result in up to ten replicasets for each deployment)

We should also verify what a failed rollout looks like, and if that's caught as a workspace failure or handled in some way by DWO.

Comment (AObuchow) [reply pkg/provision/workspace/deployment.go:235]: > if we're setting them directly it may be worth considering whether these values are appropriate for our use case. 10 minutes for progress is a very long time, and we don't necessarily need 10 revisions in the history (and I believe this setting would result in up to ten replicasets for each deployment)

Agreed. I initially added them because I thought they were necessary to prevent issues with the spec/cluster deployment diff, but they actually aren't. However, it's worth considering whether we should set (sane) values instead of letting them be the defaults set by Kubernetes. 

> We should also verify what a failed rollout looks like, and if that's caught as a workspace failure or handled in some way by DWO.

Agreed, will look into it.

**Comment by AObuchow** [comment]
@amisevsk another issue I've encountered is fixing the bug W.R.T the workspace status saying it's ready before the rolling update has completed.  I don't think the [deployment status struct fields](https://github.com/devfile/devworkspace-operator/blob/main/pkg/library/status/check.go#L57) have enough granularity to determine when the rolling update has finished (I tried something like `deployment.Status.ReadyReplicas > 0 && deployment.Status.UnavailableReplicas == 0` and other variations). 

I think I'll have to add some extra logic for checking the RollingUpdate deployment status specifically when it's used. It's a bit tricky to debug this properly since the RollingDeployment update progresses quickly :P

**Comment by AObuchow** [comment]
> @amisevsk another issue I've encountered is fixing the bug W.R.T the workspace status saying it's ready before the rolling update has completed. I don't think the [deployment status struct fields](https://github.com/devfile/devworkspace-operator/blob/main/pkg/library/status/check.go?rgh-link-date=2023-03-03T18%3A19%3A42Z#L57) have enough granularity to determine when the rolling update has finished (I tried something like `deployment.Status.ReadyReplicas > 0 && deployment.Status.UnavailableReplicas == 0` and other variations).
> 
> I think I'll have to add some extra logic for checking the RollingUpdate deployment status specifically when it's used. It's a bit tricky to debug this properly since the RollingDeployment update progresses quickly :P

This has now been addressed in [`checkRollingUpdateDeploymentStatus()`](https://github.com/devfile/devworkspace-operator/pull/1057/files#diff-2e34131f532ed47948e0f42496cfa1ef129c9f60d040ca0a80ce568cecd965a4R66)

File: pkg/library/status/check.go [hunk 2]
```diff
@@ -51,12 +51,34 @@ var unrecoverablePodEventReasons = map[string]int32{
 
 var unrecoverableDeploymentConditionReasons = []string{
 	"FailedCreate",
+	"ProgressDeadlineExceeded",
```
Comment (AObuchow) [inline pkg/library/status/check.go:54]: @amisevsk if we catch the `ProgressDeadlineExceeded` condition as I'm doing here, we should probably also specify `ProgressDeadlineSeconds` explicitly in the deployment spec (as 10 minutes is a long time to wait to see if the deployment rollout failed). Maybe `config.workspace.progressTimeout` would be a reasonable parameter to use for this?

File: pkg/provision/workspace/deployment.go (hunk 1 above)
Comment (AObuchow) [reply pkg/provision/workspace/deployment.go]: > We should also verify what a failed rollout looks like, and if that's caught as a workspace failure or handled in some way by DWO.

Not sure if this is sufficient of a check, but I checked the case you mentioned [here](https://github.com/eclipse/che/issues/21974#issuecomment-1453826209), and after modifying the memoryRequest and memoryLimit to 9Gi, the DevWorkspace failed:

```
[aobuchow@fedora devworkspace-operator]$ kubectl get dw -n $NAMESPACE -w
NAME                 DEVWORKSPACE ID             PHASE      INFO
code-latest          workspacefcb80b330a494679   Starting   Initializing DevWorkspace
code-latest          workspacefcb80b330a494679   Starting   Storage ready
code-latest          workspacefcb80b330a494679   Starting   Preparing networking
code-latest          workspacefcb80b330a494679   Starting   Preparing services
code-latest          workspacefcb80b330a494679   Starting   Preparing services
code-latest          workspacefcb80b330a494679   Starting   Networking ready
code-latest          workspacefcb80b330a494679   Starting   Waiting for DevWorkspace ServiceAccount
code-latest          workspacefcb80b330a494679   Starting   Waiting for DevWorkspace pull secrets
code-latest          workspacefcb80b330a494679   Starting   Waiting for workspace deployment
code-latest          workspacefcb80b330a494679   Starting   Waiting for editor to start
code-latest          workspacefcb80b330a494679   Running    https://workspacefcb80b330a494679.apps.ci-ln-ch9m0rb-72292.origin-ci-int-gce.dev.rhcloud.com/che-code/

code-latest          workspacefcb80b330a494679   Starting   Networking ready
code-latest          workspacefcb80b330a494679   Starting   Waiting for workspace deployment
code-latest          workspacefcb80b330a494679   Failing    Detected unrecoverable event FailedScheduling: 0/6 nodes are available: 1 Insufficient memory, 2 node(s) had volume node affinity conflict, 3 node(s) had untolerated taint {node-role.kubernetes.io/master: }. preemption: 0/6 nodes are available: 1 No preemption victims found for incoming pod, 5 Preemption is not helpful for scheduling..
```

I still haven't tested getting the deployment rollout to not complete before  `ProgressDeadlineSeconds`

File: pkg/provision/storage/shared.go [hunk 3]
```diff
@@ -198,11 +198,11 @@ func getWorkspaceVolumes(workspace *common.DevWorkspaceWithConfig) (persistent,
...
-// 1. The projects volume is not needed for the workspace (no component has mountSources: true)
-// 2. The projects volume is needed but not defined in the devfile. This is the usual case, as the projects volume
-//    is implicitly defined by mountSources
-// 3. The projects volume is explicitly defined in the workspace, as a regular volume.
-// 4. The projects volume is explicitly defined in the workspace as an ephemeral volume.
+//  1. The projects volume is not needed for the workspace (no component has mountSources: true)
+//  2. The projects volume is needed but not defined in the devfile. This is the usual case, as the projects volume
+//     is implicitly defined by mountSources
+//  3. The projects volume is explicitly defined in the workspace, as a regular volume.
+//  4. The projects volume is explicitly defined in the workspace as an ephemeral volume.
```
Comment (amisevsk) [inline pkg/provision/storage/shared.go:205]: Not sure where these whitespace changes are coming from

File: apis/controller/v1alpha1/doc.go
[the remark this answers is not in the document]
Comment (AObuchow) [reply apis/controller/v1alpha1/doc.go:1]: Yeah, I gotta fix that, something weird is going on :facepalm: VSCode is changing the license headers, and the whitespace added to other files resulted from either `go fmt` or `make fmt`.

File: pkg/library/status/check.go [hunk 4]
```diff
@@ -51,12 +51,34 @@ var unrecoverablePodEventReasons = map[string]int32{
...
+		if deployment.Status.Replicas > deployment.Status.UpdatedReplicas {
+			return false
+		}
+		if deployment.Status.AvailableReplicas < deployment.Status.UpdatedReplicas {
+			return false
+		}
+		return true
+	}
+	return false
+}
```
Comment (amisevsk) [inline pkg/library/status/check.go:80]: I believe this check can be simplified, as the kubectl implementation is designed to report specific information about the state of the rollout. Is it sufficient to check `updatedReplicas == availableReplicas == replicas == 1`?

```suggestion
func checkRollingUpdateDeploymentStatus(deployment *appsv1.Deployment) (ready bool) {
	if deployment.Generation > deployment.Status.ObservedGeneration {
		// Current deployment spec not observed by cluster
		return false
	}
	if deployment.Status.UpdatedReplicas < 1 {
		// Replica has not been updated
		return false
	}
	if deployment.Status.AvailableReplicas < 1 {
		// Updated replica is not available yet
		return false
	}
	if deployment.Status.Replicas != 1 {
		// Old replica is still terminating
		return false // Or true -- the new replica is indeed running and can be accessed
		             // why wait for the old one to terminate before letting the user see
								 // their workspace
	}
	return true
}
```

In particular, we don't really care if the deployment is completed from the cluster's perspective. All we care about is that the updated replica is available (we don't need to wait for termination of the old one necessarily, it _should_ be caught by events if it times out.

File: pkg/provision/workspace/deployment.go [hunk 5]
```diff
@@ -300,6 +315,23 @@ func getSpecDeployment(
...
+// Returns the ProgressDeadlineSeconds to use for workspace deployments as an int32.
+// The ProgressDeadLineSeconds returned is the same length of time as DWOC's workspace.ProgressTimeout.
+// Returns an error if ProgressTimeout could not be properly parsed,
+// or if the ProgressTimeout would exceed the maximum capacity of an int32
+func getProgressDeadlineSeconds(config *v1alpha1.OperatorConfiguration) (int32, error) {
+	timeout, err := time.ParseDuration(config.Workspace.ProgressTimeout)
+	if err != nil {
+		return 0, fmt.Errorf("invalid duration specified for workspace progress timeout: %w", err)
+	}
+	timeoutRounded := timeout.Round(time.Second)
```
Comment (amisevsk) [inline pkg/provision/workspace/deployment.go:327]: Do you need to round? I believe converting to an `int32` will drop the factional part.

**Review by amisevsk — COMMENTED** [review]
Looking good, just a couple more changes.

File: pkg/provision/workspace/deployment.go (hunk 5 above)
Comment (AObuchow) [reply pkg/provision/workspace/deployment.go:327]: You're right, no need to round -- I was just being overly cautious :P

File: pkg/provision/workspace/deployment.go (hunk 1 above)
Comment (AObuchow) [reply pkg/provision/workspace/deployment.go]: > we don't necessarily need 10 revisions in the history (and I believe this setting would result in up to ten replicasets for each deployment)

Last thing I thought about for this PR: I checked the deployment spec and the default is indeed 10 for the revision history limit. Maybe this should be modified to 2? Or even 3? Or we can leave this for another PR/if an issue arises from this in the future.

File: pkg/library/status/check.go (hunk 4 above)
Comment (AObuchow) [reply pkg/library/status/check.go]: +1 That's a lot better, thanks for the suggestion :)

**Review by amisevsk — APPROVED** [review]
LGTM
